File: macos/Sources/BambuDryCore/Control/AutoDryController.swift
import Foundation

/// Hysteresis state machine for one AMS unit.
///
/// Behaviour:
/// - Issue START when humidity ≥ `highThreshold`, AMS isn't already drying,
///   nothing in `cannotDryReasons`, anti-cycling time elapsed, and we haven't
///   already sent a START in the last `commandRetryInterval`.
/// - Issue STOP when humidity ≤ `lowThreshold`, AMS is currently drying,
///   `minOnTime` has elapsed since the last START, and we haven't already
///   sent a STOP in the last `commandRetryInterval`.
/// - Skip everything if `runDuringPrint` is false and the printer is printing.
///
/// The "in-flight retry" window means: if a START is published but the printer
/// hasn't begun reporting `isCurrentlyDrying = true` yet, we won't spam more
/// STARTs every 5s. After the retry window we *will* re-issue, in case the
/// command was lost.
public actor AutoDryController {

    public enum Decision: Sendable, Equatable {
        case start(amsId: Int, filament: String, tempC: Int, hours: Int)
        case stop(amsId: Int)
        case noop(reason: NoopReason)
    }

    public enum NoopReason: Sendable, Equatable {
        case disabled
        case unsupportedAMS
        case humidityUnknown
        case printing
        case withinHysteresisBand
        /// Anti-cycling: we just stopped, not allowed to start again yet.
        /// `secondsRemaining` counts down to when we'd permit a start.
        case waitingAfterStop(secondsRemaining: Int)
        /// Anti-cycling: we just started, not allowed to stop again yet.
        case waitingAfterStart(secondsRemaining: Int)
        /// We sent a START recently and are waiting for the printer to ack.
        case startInFlight(secondsRemaining: Int)
        /// We sent a STOP recently and are waiting for the printer to ack.
        case stopInFlight(secondsRemaining: Int)
        case cannotDry([CannotDryReason])
        case settingsInvalid
    }

    public let amsId: Int
    /// Re-issue same command if this elapses without the printer acking it.
    /// Not user-facing; protects against lost commands.
    public let commandRetryInterval: TimeInterval
    public let dryHours: Int

    private var lastStartIssuedAt: Date
    private var lastStopIssuedAt: Date
    private let clock: @Sendable () -> Date

    public init(amsId: Int,
                commandRetryInterval: TimeInterval = 30,
                dryHours: Int = 1,
                clock: @Sendable @escaping () -> Date = { Date() }) {
        self.amsId = amsId
        self.commandRetryInterval = commandRetryInterval
        self.dryHours = dryHours
        self.clock = clock
        self.lastStartIssuedAt = .distantPast
        self.lastStopIssuedAt = .distantPast
    }

    public func evaluate(snapshot: AMSSnapshot,
                         settings: AutoDrySettings,
                         printerIsPrinting: Bool) -> Decision {

        guard settings.enabled else { return .noop(reason: .disabled) }
        guard settings.isValid else { return .noop(reason: .settingsInvalid) }
        guard snapshot.supportsDrying else { return .noop(reason: .unsupportedAMS) }
        guard let rh = snapshot.humidityPercent else { return .noop(reason: .humidityUnknown) }
        if printerIsPrinting && !settings.runDuringPrint { return .noop(reason: .printing) }

        let now = clock()
        let minOnTime  = TimeInterval(settings.minOnMinutes  * 60)
        let minOffTime = TimeInterval(settings.minOffMinutes * 60)

        // If the AMS reports it's already drying, the printer has acted on (or
        // beat us to) a start command — anchor lastStart to now so future
        // stop-time gating works from the actual drying-start moment.
        if snapshot.isCurrentlyDrying && lastStartIssuedAt == .distantPast {
            lastStartIssuedAt = now
        }

        // Rising edge: humidity high, AMS not drying → consider START
        if rh >= settings.highThreshold, !snapshot.isCurrentlyDrying {
            if !snapshot.cannotDryReasons.isEmpty {
                return .noop(reason: .cannotDry(snapshot.cannotDryReasons))
            }
            // Don't restart immediately after a stop (anti-cycling)
            let sinceStop = now.timeIntervalSince(lastStopIssuedAt)
            if sinceStop < minOffTime {
                return .noop(reason: .waitingAfterStop(secondsRemaining: Int(ceil(minOffTime - sinceStop))))
            }
            // If we already issued a START recently, wait for printer to ack
            // (or for retry window to elapse, in case the command was lost).
            let sinceStart = now.timeIntervalSince(lastStartIssuedAt)
            if sinceStart < commandRetryInterval {
                return .noop(reason: .startInFlight(secondsRemaining: Int(ceil(commandRetryInterval - sinceStart))))
            }
            lastStartIssuedAt = now
            return .start(amsId: amsId,
                          filament: snapshot.dominantFilamentType ?? "PLA",
                          tempC: settings.targetTemp,
                          hours: dryHours)
        }

        // Falling edge: humidity low, AMS drying → consider STOP
        if rh <= settings.lowThreshold, snapshot.isCurrentlyDrying {
            // Don't stop immediately after a start (let it actually dry)
            let sinceStart = now.timeIntervalSince(lastStartIssuedAt)
            if sinceStart < minOnTime {
                return .noop(reason: .waitingAfterStart(secondsRemaining: Int(ceil(minOnTime - sinceStart))))
            }
            // Don't re-issue STOP if we just sent one
            let sinceStop = now.timeIntervalSince(lastStopIssuedAt)
            if sinceStop < commandRetryInterval {
                return .noop(reason: .stopInFlight(secondsRemaining: Int(ceil(commandRetryInterval - sinceStop))))
            }
            lastStopIssuedAt = now
            return .stop(amsId: amsId)
        }

        return .noop(reason: .withinHysteresisBand)
    }

    // MARK: - Test / persistence helpers

    public func setLastStartIssuedAt(_ date: Date) { lastStartIssuedAt = date }
    public func setLastStopIssuedAt(_ date: Date)  { lastStopIssuedAt = date }
    public func getLastStartIssuedAt() -> Date { lastStartIssuedAt }
    public func getLastStopIssuedAt() -> Date { lastStopIssuedAt }
}

File: windows/src/BambuDry.Core/Control/AutoDryController.cs
namespace BambuDry.Core.Control;

public abstract record Decision
{
    public sealed record Start(int AmsId, string Filament, int TempC, int Hours) : Decision;
    public sealed record Stop(int AmsId) : Decision;
    public sealed record Noop(NoopReason Reason) : Decision;
}

public abstract record NoopReason
{
    public sealed record Disabled : NoopReason;
    public sealed record UnsupportedAms : NoopReason;
    public sealed record HumidityUnknown : NoopReason;
    public sealed record Printing : NoopReason;
    public sealed record WithinHysteresisBand : NoopReason;
    /// <summary>Anti-cycling: just stopped, can't start yet.</summary>
    public sealed record WaitingAfterStop(int SecondsRemaining) : NoopReason;
    /// <summary>Anti-cycling: just started, can't stop yet.</summary>
    public sealed record WaitingAfterStart(int SecondsRemaining) : NoopReason;
    /// <summary>Sent a START recently; waiting for printer to ack.</summary>
    public sealed record StartInFlight(int SecondsRemaining) : NoopReason;
    /// <summary>Sent a STOP recently; waiting for printer to ack.</summary>
    public sealed record StopInFlight(int SecondsRemaining) : NoopReason;
    public sealed record SettingsInvalid : NoopReason;

    public sealed record CannotDry(IReadOnlyList<CannotDryReason> Reasons) : NoopReason
    {
        public bool Equals(CannotDry? other) =>
            other is not null && Reasons.SequenceEqual(other.Reasons);

        public override int GetHashCode()
        {
            var hc = new HashCode();
            foreach (var r in Reasons) hc.Add(r);
            return hc.ToHashCode();
        }
    }
}

/// <summary>
/// Hysteresis state machine for one AMS unit.
///
/// Behaviour:
/// - Issue START when humidity ≥ HighThreshold, AMS isn't already drying,
///   no CannotDryReasons, anti-cycling time elapsed, no recent in-flight START.
/// - Issue STOP when humidity ≤ LowThreshold, AMS is currently drying,
///   MinOnTime elapsed since the last START, no recent in-flight STOP.
/// - Skip everything if RunDuringPrint is false and the printer is printing.
/// </summary>
public sealed class AutoDryController
{
    public int AmsId { get; }
    /// <summary>Re-issue same command if this elapses without the printer acking it.</summary>
    public TimeSpan CommandRetryInterval { get; }
    public int DryHours { get; }

    private readonly Func<DateTime> _clock;
    private readonly object _lock = new();
    private DateTime _lastStartIssuedAt = DateTime.MinValue;
    private DateTime _lastStopIssuedAt  = DateTime.MinValue;

    public AutoDryController(int amsId, TimeSpan? commandRetryInterval = null, int dryHours = 1, Func<DateTime>? clock = null)
    {
        AmsId = amsId;
        CommandRetryInterval = commandRetryInterval ?? TimeSpan.FromSeconds(30);
        DryHours = dryHours;
        _clock = clock ?? (() => DateTime.UtcNow);
    }

    public Decision Evaluate(AmsSnapshot snapshot, AutoDrySettings settings, bool printerIsPrinting)
    {
        if (!settings.Enabled)         return new Decision.Noop(new NoopReason.Disabled());
        if (!settings.IsValid)         return new Decision.Noop(new NoopReason.SettingsInvalid());
        if (!snapshot.SupportsDrying)  return new Decision.Noop(new NoopReason.UnsupportedAms());
        if (snapshot.HumidityPercent is not int rh) return new Decision.Noop(new NoopReason.HumidityUnknown());
        if (printerIsPrinting && !settings.RunDuringPrint) return new Decision.Noop(new NoopReason.Printing());

        lock (_lock)
        {
            var now        = _clock();
            var minOnTime  = TimeSpan.FromMinutes(settings.MinOnMinutes);
            var minOffTime = TimeSpan.FromMinutes(settings.MinOffMinutes);

            // If the AMS reports it's already drying, anchor lastStart to now so future
            // stop-time gating works from the actual drying-start moment.
            if (snapshot.IsCurrentlyDrying && _lastStartIssuedAt == DateTime.MinValue)
                _lastStartIssuedAt = now;

            // Rising edge: humidity high, AMS not drying → consider START
            if (rh >= settings.HighThreshold && !snapshot.IsCurrentlyDrying)
            {
                if (snapshot.CannotDryReasons.Count > 0)
                    return new Decision.Noop(new NoopReason.CannotDry(snapshot.CannotDryReasons));

                var sinceStop = now - _lastStopIssuedAt;
                if (sinceStop < minOffTime)
                    return new Decision.Noop(new NoopReason.WaitingAfterStop(CeilSeconds(minOffTime - sinceStop)));

                var sinceStart = now - _lastStartIssuedAt;
                if (sinceStart < CommandRetryInterval)
                    return new Decision.Noop(new NoopReason.StartInFlight(CeilSeconds(CommandRetryInterval - sinceStart)));

                _lastStartIssuedAt = now;
                return new Decision.Start(
                    AmsId: AmsId,
                    Filament: snapshot.DominantFilamentType ?? "PLA",
                    TempC: settings.TargetTemp,
                    Hours: DryHours);
            }

            // Falling edge: humidity low, AMS drying → consider STOP
            if (rh <= settings.LowThreshold && snapshot.IsCurrentlyDrying)
            {
                var sinceStart = now - _lastStartIssuedAt;
                if (sinceStart < minOnTime)
                    return new Decision.Noop(new NoopReason.WaitingAfterStart(CeilSeconds(minOnTime - sinceStart)));

                var sinceStop = now - _lastStopIssuedAt;
                if (sinceStop < CommandRetryInterval)
                    return new Decision.Noop(new NoopReason.StopInFlight(CeilSeconds(CommandRetryInterval - sinceStop)));

                _lastStopIssuedAt = now;
                return new Decision.Stop(AmsId);
            }

            return new Decision.Noop(new NoopReason.WithinHysteresisBand());
        }
    }

    private static int CeilSeconds(TimeSpan t) => (int)Math.Ceiling(t.TotalSeconds);

    // Test / persistence helpers
    public void SetLastStartIssuedAt(DateTime t) { lock (_lock) _lastStartIssuedAt = t; }
    public void SetLastStopIssuedAt(DateTime t)  { lock (_lock) _lastStopIssuedAt = t; }
    public DateTime GetLastStartIssuedAt() { lock (_lock) return _lastStartIssuedAt; }
    public DateTime GetLastStopIssuedAt()  { lock (_lock) return _lastStopIssuedAt; }
}
